_posts/2017-10-04-scalamacros.md

```diff
@@ -0,0 +1,299 @@
+establishing a standard, non-experimental macro system for Scala.
+I'll try to keep it short here, for a more comprehensive coverage please refer
+to the list of [papers written by Eugene Burmako][papers], the founder of Scala
+Macros, in particular his PhD thesis [Unification of Compile-Time and Runtime
+Metaprogramming in Scala][thesis].
+
+If you want the TL;DR, see [next steps](#next-steps).
+
+### v1: scala.reflect
+
+Scala.reflect-based macros are an integral part of the Scala 2.x ecosystem.
+Well-known libraries like ScalaTest, Sbt, Spark, Circe, Slick, Shapeless,
+Spire and others use scala.reflect macros to achieve previously unreachable
+standards of expressiveness, type safety and performance.
+
+Unfortunately, scala.reflect-based macros have also gained notoriety as an
+arcane and brittle technology.
+The most common criticisms of scala.reflect is its sub-par tooling support
+and non-portable metaprogramming API based on Scala 2.x compiler internals.
+Even five years after their introduction, scala.reflect macros still can't
+expand in IntelliJ, leading to proliferation of spurious red squiggles -
+sometimes in pretty simple code.
+
+Quoting Eugene Burmako, the author of Scala macros, in [SIP-16] on
+self-cleaning macros
+
+> While trying to fix these problems via evolutionary changes to the current
+> macro system, we have realized that most of them are caused by the decision
+> to use scala.reflect as the underlying metaprogramming API. Modelled after
+> compiler internals, scala.reflect inherits many of its peculiar design
+> choices. Extensive use of desugarings and existence of multiple independent
+> program representations may have worked well for compiler development, but
+> they turned out to be inadequate for a public API.
+
+As a result of these known limitations, the language committee has decided to
+retire the scala.reflect-based macro system.
+Another justification for retiring the scala.reflect-based macro system was
+that a new macro system based on Scalameta was "just around the corner".
+
+### v2: scala.meta
+
+The [Scalameta] project was founded to become a better macro system for Scala,
+with the vision to replace scala.reflect as the de-facto metaprogramming
+toolkit for Scala.
+
+With Scalameta, we managed to support macro annotations in
+[Scala 2.x](http://scalameta.org/paradise/),
+[IntelliJ Scala plugin](https://blog.jetbrains.com/scala/2016/11/11/intellij-idea-2016-3-rc-scala-js-scala-meta-and-more/)
+and
+[Dotty](https://github.com/liufengyun/eden).
+The novelty with Scalameta macros was that they converted compiler-specific
+ASTs into the [Scalameta AST], which is a large collection "dumb" data
+containers that leak no implementation details from the compiler.
+The details of this approach are explained in more detail in [SIP-29], a
+proposal to use Scalameta as the foundation for building macros in Scala.
+
+In Scalafmt, I use Scalameta macro annotations to generate readers for
```

Comment:
Probably should expand here too... Context would be helpful? I think this is too concise for most people.

Reply:
Updated.